spec/features/saml/saml_spec.rb

```diff
@@ -273,6 +279,240 @@
+        expect(user.active_identities.size).to eq(0)
+      end
+
+      it 'asserts correct SessionIndex' do
+        expect(request_xmldoc.asserted_session_index).
+          to eq(@sp1_asserted_session_index)
+        click_button 'Submit' # LogoutRequest for first SP
+
+        # SP1 logs user out and responds with success
+        # User session is terminated at IdP and success
+        # is returned to SP2 (originating requestor)
+        expect(response_xmldoc.logout_status_assertion).
+          to eq('urn:oasis:names:tc:SAML:2.0:status:Success')
+
+        click_button 'Submit' # LogoutResponse for originating SP
+
+        visit edit_user_registration_path
+        expect(page).to have_content I18n.t 'devise.failure.unauthenticated'
+      end
+
+      it 'terminates the user session at IdP' do
+        click_button 'Submit' # LogoutRequest for first SP
+        click_button 'Submit' # LogoutResponse for originating SP
+
+        visit edit_user_registration_path
+        expect(page).to have_content I18n.t 'devise.failure.unauthenticated'
+      end
+    end
+
+    context 'with multiple SP sessions and SP-initiated logout; non-chronological' do
+      let(:user) { create(:user, :signed_up) }
+      let(:response_xmldoc) { XmlDoc.new('feature', 'response_assertion') }
+      let(:request_xmldoc) { XmlDoc.new('feature', 'request_assertion') }
+
+      before do
+        visit sp1_authnrequest # sp1
+        authenticate_user(user)
+        @sp1_session_index = response_xmldoc.response_session_index_assertion
+        click_button 'Submit'
+
+        visit sp2_authnrequest # sp2
+        @sp2_session_index = response_xmldoc.response_session_index_assertion
+        click_button 'Submit'
+
+        request = OneLogin::RubySaml::Logoutrequest.new
+        settings = sp1_saml_settings
+        settings.name_identifier_value = user.uuid
+        visit request.create(settings) # sp1
+      end
+
+      it 'terminates the user session at IdP' do
+        click_button 'Submit' # LogoutRequest for first SP: sp2
+        click_button 'Submit' # LogoutResponse for originating SP: sp1
+
+        visit edit_user_registration_path
+        expect(page).to have_content I18n.t 'devise.failure.unauthenticated'
+      end
+
+      it 'terminates sessions in order of authentication' do
+        expect(request_xmldoc.asserted_session_index).to eq(@sp2_session_index)
+        click_button 'Submit' # LogoutRequest for first SP: sp2
+
+        expect(response_xmldoc.logout_status_assertion).
+          to eq('urn:oasis:names:tc:SAML:2.0:status:Success')
+        click_button 'Submit' # LogoutResponse for originating SP: sp1
+
+        visit edit_user_registration_path
+        expect(page).to have_content I18n.t 'devise.failure.unauthenticated'
+
+        expect(user.active_identities.size).to eq(0)
+      end
+    end
+
+    context 'with alternate multiple SP sessions and SP-initiated logout; non-chronological' do
```

Comment:
This context looks like a duplicate of the one above with the sp1 and sp2 swapped.  Maybe it can be re-factored to use shared examples or some other way to avoid the duplication?